### PR TITLE
Upgrade to 2.7.4

### PR DESCRIPTION
**Pull request by yalh76**
## Problem
- *Update to 2.7.4*

## Solution
- *Apply example_ynh + french and english readme*
- *Use of experimental helper ynh_install_ruby*
- *Add custom .env.production.sample*
- *Fix remove / backup / restore / upgrade scripts*
- *Implement change_url*

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Validation
---
- [ ] **Code review**
- [ ] **Approval (LGTM)** 
- **CI succeeded** : 
[Build Status/badge/icon)](https://ci-apps-dev.yunohost.org/jenkins/job/mastodon_ynh_PR130%20(yalh76)/)

Revision: 0494ace9aff9eaa8395bc0187f140c0eb975e645 
Capture

### CODE REVIEW

**Comment by yalh76** [comment]
Sorry @YunoHost-Apps/apps-group a lot of changes

File: check_process [hunk 1]
```diff
@@ -15,13 +14,9 @@
...
 		backup_restore=1
 		multi_instance=0
-		wrong_user=0
-		wrong_path=0
 		incorrect_path=0
-		corrupt_source=0
-		fail_download_source=0
 		port_already_use=0
-		final_path_already_use=0
+		change_url=0
```
Comment (maniackcrudelis) [inline check_process]: As you add a change_url script, this test should be set at 1

File: scripts/install [hunk 2]
```diff
@@ -8,35 +8,44 @@
...
-source _future.sh
+source ynh_install_ruby
 
 #=================================================
 # MANAGE SCRIPT FAILURE
 #=================================================
 
+ynh_clean_setup () {
+	### Remove this function if there's nothing to clean before calling the remove script.
+	read -p "Press any key..."
```
Comment (maniackcrudelis) [inline scripts/install]: ?

File: scripts/install [hunk 3]
```diff
@@ -8,35 +8,44 @@
...
 #=================================================
 # RETRIEVE ARGUMENTS FROM THE MANIFEST
-#================================================
+#=================================================
 
 domain=$YNH_APP_ARG_DOMAIN
+path_url="/"
 admin_mastodon=$YNH_APP_ARG_ADMIN
-admin_mastodon_mail=$(ynh_user_get_info $admin_mastodon 'mail')
+is_public=true
```
Comment (maniackcrudelis) [inline scripts/install]: Why do you keep that variable ?
This parameter isn't asked to the admin.

File: scripts/install (hunk 2 above)
Comment (yalh76) [reply scripts/install]: lol testing purpose ^^, during packaging....

File: scripts/upgrade [hunk 4]
```diff
@@ -131,68 +113,88 @@ ynh_secure_remove $final_path/live/config/initializers/timeout.rb
...
+#=================================================
+ynh_print_info "Upgrading dependencies..."
+
+# Install source.list debian package backports & yarn
+if [ "$(lsb_release --codename --short)" == "jessie" ]; then
+	echo "deb http://httpredir.debian.org/debian jessie-backports main" | tee /etc/apt/sources.list.d/jessie-backports.list
+fi
+curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
+echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+ynh_package_update
```
Comment (maniackcrudelis) [inline scripts/upgrade]: Please have a look to this experimental helper to handle extra repo
https://github.com/YunoHost-Apps/Experimental_helpers/blob/master/ynh_add_secure_repos/ynh_add_secure_repos__2

File: scripts/install (hunk 3 above)
Comment (yalh76) [reply scripts/install]: in the original install script @nemsia wrote:
`ODO: all private install`
https://github.com/YunoHost-Apps/mastodon_ynh/blob/master/scripts/install#L279

I do the first step asI don't know if it's agood idea or not to install a private mastodon and I havan't look for now how to disable mastodon federation for a private instance

**Comment by maniackcrudelis** [comment]
That makes really a lot of change for an app a little bit difficult to handle I think.
I hope all of this is fully tested.

File: scripts/upgrade (hunk 4 above)
Comment (yalh76) [reply scripts/upgrade]: Cool, I haven't seen it, I add to watch Experimantal_helper

File: scripts/install (hunk 3 above)
Comment (maniackcrudelis) [reply scripts/install]: ok
In that case, add a comment to explain it. Otherwise, others may ask the same question.

**Comment by yalh76** [comment]
> 
> 
> That makes really a lot of change for an app a little bit difficult to handle I think.
> I hope all of this is fully tested.

Well, I haven't' tested Mastodon functionalities, just done a package_check and installed it
Capture

**Comment by maniackcrudelis** [comment]
Ok, that's already quite good.
Maybe you could try a public testing as soon as it's ready to merge.
However, I don't know if there's many people using this package.

**Comment by yalh76** [comment]
> 
> 
> Ok, that's already quite good.
> Maybe you could try a public testing as soon as it's ready to merge.
> However, I don't know if there's many people using this package.

I will first try to :
- [ ] implement private instance
- [ ] implement LDAP support
- [ ] Example_ynh scripts cleanup

Seems that there are some users: 
- complain of an old upstream version: https://github.com/YunoHost-Apps/mastodon_ynh/issues/128
- install fail : https://github.com/YunoHost-Apps/mastodon_ynh/issues/131

**Comment by maniackcrudelis** [comment]
I think you should keep LDAP and private instance for later.
Try first to release a working version, up to date.
It will be easier for you to work on it after.

File: scripts/upgrade (hunk 4 above)
Comment (yalh76) [reply scripts/upgrade]: extra repo handled

Comment (maniackcrudelis) [reply scripts/upgrade]: :+1:

**Comment by yalh76** [comment]
Even if working, for now package_check stuck by a bug with ynh_add_secure_repos__2: YunoHost-Apps/Experimental_helpers#36

**Comment by anmol26s** [comment]
@yalh76 
When the PR is ready ping me as it has become very big. Its better to test when everything is done.
Do mention the major changes done, so I can test them(eg. Ldap,change_url, etc).

Thanks for your love and work for the packages.

**Comment by yalh76** [comment]
@anmol26s PR is ready for testing.
Major changes: 

- Update to Mastodon 2.7.4
- Fix remove / backup / restore
- Implement change_url
- Implement public/private

There is no LDAP support for now.

**Comment by anmol26s** [comment]
> Implement change_url
But is it possible to change the instance url once installed for mastodon? I don't know if its good idea as this may break the instance federation capabilities.

**Comment by yalh76** [comment]
> 
> 
> > Implement change_url
> 
> But is it possible to change the instance url once installed for mastodon? I don't know if its good idea as this may break the instance federation capabilities.
> 
> Here is the discussion: https://discourse.joinmastodon.org/t/domain-changes-and-aliases/671/4

I agree, changing the url will break the federation. for other federated instances, it will be like a new instance

**Comment by anmol26s** [comment]
So should we add this in readme or remove the change_url option? For other networks like Friendica,Hubzilla and Peertube, I don't prefer this because they have there own way to migrate to new instances. And if the change_url will work as new instance, there is no point in migrate to new url, when a new instance can be installed through multi-instance.
I would like to take suggestion on this @YunoHost-Apps/apps-group and @nemsia 

@yalh76 
Have you tested the Ldap configuration ? This is most requested feature for the this app. I have played around with LDAP in past, but it do not work. Is the any scope of making LDAP work in this PR itself ? People will like this feature with working app.

**Comment by yalh76** [comment]
@anmol26s 

Yes I tried to activate LDAP: https://github.com/YunoHost-Apps/mastodon_ynh/pull/130/commits/37e462d1adacc1f011aedac4209bccae36535275

But Mastodon request a user and password to access to LDAP and it's recommanded to use a dedicated account.
https://github.com/tootsuite/documentation/blob/master/Running-Mastodon/Enabling-LDAP-login.md
`When Mastodon needs to search for a user account in your LDAP server, it will first authenticate (bind) to the server using these credentials to do the search. It's recommended to create a service account for Mastodon in your LDAP directory, and give it a strong password. Then use the DN (Distinguished Name) and password of the service account for these settings.`

As there isn't LDAP helper I wasn't able to create a dedicated LDAP account during the installation so I removed my attempt to implement LDAP

**Comment by anmol26s** [comment]
2.8 is out
https://blog.joinmastodon.org/2019/04/mastodon-2.8/

**Comment by yalh76** [comment]
Maybe we can first validate this PR, master is far beyond the version include.
When merged, upgrade to 2.8.0 will took 2 minutes...

**Comment by anmol26s** [comment]
So, what should be done for the change_url? My vote is to remove it.

**Comment by yalh76** [comment]
> 
> 
> So, what should be done for the change_url? My vote is to remove it.

I made a Poll in https://github.com/YunoHost-Apps/mastodon_ynh/issues/133, it request Pollsto be installed in mastodon_ynh ^^

**Comment by anmol26s** [comment]
@yalh76 
Why you have removed this PR? https://github.com/YunoHost-Apps/mastodon_ynh/pull/129

**Comment by anmol26s** [comment]
This should be removed or changed to 2.7.4: https://github.com/YunoHost-Apps/mastodon_ynh/pull/130/commits/8bcf44150e7ea5a65efe536179895e95dc863217#diff-04c6e90faac2675aa89e2176d2eec7d8R11

This should be removed or changed from testing to working:
https://github.com/YunoHost-Apps/mastodon_ynh/pull/130/commits/8bcf44150e7ea5a65efe536179895e95dc863217#diff-04c6e90faac2675aa89e2176d2eec7d8R12


This should be removed or changed to the version listed in manifest:
https://github.com/YunoHost-Apps/mastodon_ynh/pull/130/commits/8bcf44150e7ea5a65efe536179895e95dc863217#diff-04c6e90faac2675aa89e2176d2eec7d8R15

For me all these lines from 11 to 16 should be removed.

**Comment by yalh76** [comment]
> 
> 
> @yalh76
> Why you have removed this PR? #129

I didn't removed  PR #129
I forked the project before you [merged it](https://github.com/YunoHost-Apps/mastodon_ynh/network):
Sans titre

**Comment by yalh76** [comment]
> 
> 
> This should be removed or changed to 2.7.4: [8bcf441#diff-04c6e90faac2675aa89e2176d2eec7d8R11](https://github.com/YunoHost-Apps/mastodon_ynh/commit/8bcf44150e7ea5a65efe536179895e95dc863217#diff-04c6e90faac2675aa89e2176d2eec7d8R11)
> 
> This should be removed or changed from testing to working:
> [8bcf441#diff-04c6e90faac2675aa89e2176d2eec7d8R12](https://github.com/YunoHost-Apps/mastodon_ynh/commit/8bcf44150e7ea5a65efe536179895e95dc863217#diff-04c6e90faac2675aa89e2176d2eec7d8R12)
> 
> This should be removed or changed to the version listed in manifest:
> [8bcf441#diff-04c6e90faac2675aa89e2176d2eec7d8R15](https://github.com/YunoHost-Apps/mastodon_ynh/commit/8bcf44150e7ea5a65efe536179895e95dc863217#diff-04c6e90faac2675aa89e2176d2eec7d8R15)
> 
> For me all these lines from 11 to 16 should be removed.

All badges removed.
I did not dare to delete them thinking that it's important for the mastodon_ynh's maintainer

**Comment by anmol26s** [comment]
>  I forked the project before you merged it

Can you fetch this PR in your PR?

**Comment by yalh76** [comment]
> 
> 
> > I forked the project before you merged it
> 
> Can you fetch this PR in your PR?

Fetched, waiting package_check results

**Comment by yalh76** [comment]
> 
> 
> > I forked the project before you merged it
> 
> Can you fetch this PR in your PR?

With PR #129 package check failed... the VM doesn't have enough memory

**Comment by anmol26s** [comment]
> With PR #129 package check failed... the VM doesn't have enough memory

If @alexAubin  is ok to not to include this PR for now, then we can merge this.

**Comment by alexAubin** [comment]
Yes :wink: Or maybe @maniackcrudelis knows a trick to detect the script is being run on package check or the CI

**Comment by maniackcrudelis** [comment]
Indeed I do...

But the problem here isn't about the CI or package_check. The problem is about the amount of ram available.
An experimental helper is available to add a swap file for those situations.

**Comment by yalh76** [comment]
> 
> 
> > With PR #129 package check failed... the VM doesn't have enough memory
> 
> If @alexAubin is ok to not to include this PR for now, then we can merge this.

I made an ugly workaround to let the package_check works without PR #129

**Comment by alexAubin** [comment]
You can add swap, but I dunno if that's the right solution. Consider a user running an Olimex board or Raspberry Pi 2 or idk ...
- either the user knows what he's doing and agrees to add swap (despite the fact that this can ruin the SDcard if swapiness ain't configured properly)
- either the user doesnt know what he's doing and you simply want to abort the install before it crashes

**Comment by anmol26s** [comment]
I feel that feature of adding swap memory should be done through command-line rather then the package helper.

> I made an ugly workaround to let the package_check works without PR #129 

Lets see if package check is ok with it.

**Comment by yalh76** [comment]
> 
> Lets see if package check is ok with it.

For now https://ci-apps-dev.yunohost.org/jenkins/ is down but I update the result from my own CI.

**Comment by anmol26s** [comment]
The CI has failed: https://ci-apps-dev.yunohost.org/jenkins/job/mastodon_ynh_PR130%20(yalh76)/1/console

**Comment by yalh76** [comment]
I've seen that.

I made the same build on my own CI and updated the screenshot: The build is OK. 
http://www.ipv6proxy.net/go.php?u=https%3A%2F%2Fvm00.yalh.net%2Fjenkins%2Fview%2Fmastodon%2Fjob%2Fmastodon_yalh76_develop%2F&b=1

Maybe something specific on ci-apps-dev ?

**Comment by anmol26s** [comment]
@maniackcrudelis will give more inside to it. Sometimes my apps fails completely then on next check they rise up.

**Comment by anmol26s** [comment]
Should we merge it with CI status failing? The app can't be improved till its tested by users and issues are reported.

**Comment by yalh76** [comment]
I think that we can merge it.
The build is ok in my CI. And the error in the dev-ci is related to nodejs installation using official helper...

**Comment by anmol26s** [comment]
Ok, merging this. If there is a problem it can be fixed in testing. For me code looks good.

**Pull request merged**